File: causal-engine/causal_engine/annotate.py
"""Posts a Grafana annotation for a diagnosis, so a verdict shows up as a
marker on the same dashboard load spikes and chaos faults already do
(see loadgen/annotate.go and chaos/annotate.go for the Go-side twins)."""
from __future__ import annotations

import logging

import httpx

from causal_engine import config

logger = logging.getLogger(__name__)


def post_annotation(text: str, tags: list[str], time_ms: int) -> None:
    try:
        resp = httpx.post(
            f"{config.GRAFANA_URL}/api/annotations",
            json={"time": time_ms, "tags": tags, "text": text},
            auth=(config.GRAFANA_USER, config.GRAFANA_PASS),
            timeout=3.0,
        )
        if resp.status_code >= 300:
            logger.warning("annotation post returned %s: %s", resp.status_code, resp.text)
    except Exception as exc:
        logger.warning("annotation post failed (is Grafana reachable at %s?): %s", config.GRAFANA_URL, exc)

File: control-plane/annotate.go
package main

import (
	"bytes"
	"encoding/json"
	"log"
	"net/http"
	"time"
)

// postAnnotation drops a marker on the Grafana timeline for an executed
// action. Mirrors loadgen/annotate.go, chaos/annotate.go, and
// causal-engine's annotate.py — every event source posts its own.
func postAnnotation(baseURL, user, pass, text string, tags []string, at time.Time) {
	body, err := json.Marshal(map[string]any{
		"time": at.UnixMilli(),
		"tags": tags,
		"text": text,
	})
	if err != nil {
		return
	}
	req, err := http.NewRequest(http.MethodPost, baseURL+"/api/annotations", bytes.NewReader(body))
	if err != nil {
		log.Printf("annotation request: %v", err)
		return
	}
	req.SetBasicAuth(user, pass)
	req.Header.Set("Content-Type", "application/json")

	client := &http.Client{Timeout: 3 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		log.Printf("annotation post failed (is Grafana reachable at %s?): %v", baseURL, err)
		return
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 300 {
		log.Printf("annotation post returned %s", resp.Status)
	}
}
